File: Classes/QuadInfoCell.m
// QuadInfoCell.h
//
// TableView cells with four labels for information.


#import "QuadInfoCell.h"
#import "AppDelegate.h"

@interface QuadInfoCell ()

@property (nonatomic, strong) UIView *separatorView;

@end

@implementation QuadInfoCell
{
    UITableViewCellStateMask cellState;
    BOOL large;
}


- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier enlargeTopRightLabel:(BOOL)enlargeTopRightLabel
{
    if ((self = [super initWithStyle:style reuseIdentifier:reuseIdentifier])) {

        cellState = UITableViewCellStateDefaultMask;
        large     = enlargeTopRightLabel;

    
        _topLeftLabel                            = [[UILabel alloc] initWithFrame:CGRectZero];
        _topLeftLabel.backgroundColor            = [UIColor clearColor];
        _topLeftLabel.textColor                  = [UIColor blackColor];
        _topLeftLabel.adjustsFontSizeToFitWidth  = YES;
        _topLeftLabel.font                       = [UIFont fontWithName:@"HelveticaNeue-Light" size:22];
        _topLeftLabel.minimumScaleFactor         = 12.0/[_topLeftLabel.font pointSize];
        [self.contentView addSubview:_topLeftLabel];


        _botLeftLabel                            = [[UILabel alloc] initWithFrame:CGRectZero];
        _botLeftLabel.backgroundColor            = [UIColor clearColor];
        _botLeftLabel.textColor                  = [UIColor colorWithWhite:0.5 alpha:1.0];
        _botLeftLabel.adjustsFontSizeToFitWidth  = YES;
        _botLeftLabel.font                       = [UIFont fontWithName:@"HelveticaNeue-Light" size:15.0];
        _botLeftLabel.minimumScaleFactor         = 12.0/[_botLeftLabel.font pointSize];
        [self.contentView addSubview:_botLeftLabel];


        _topRightLabel                           = [[UILabel alloc] initWithFrame:CGRectZero];
        _topRightLabel.backgroundColor           = [UIColor clearColor];
        _topRightLabel.textColor                 = [UIColor blackColor];
        _topRightLabel.adjustsFontSizeToFitWidth = YES;
        _topRightLabel.font                      = [UIFont fontWithName:@"HelveticaNeue-Light" size:large ? 28.0 : 22.0];
        _topRightLabel.minimumScaleFactor        = 12.0/[_topRightLabel.font pointSize];
        _topRightLabel.textAlignment             = NSTextAlignmentRight;
        [self.contentView addSubview:_topRightLabel];


        _botRightLabel = [[UILabel alloc] initWithFrame:CGRectZero];
        _botRightLabel.backgroundColor           = [UIColor clearColor];
        _botRightLabel.textColor                 = [UIColor colorWithWhite:0.5 alpha:1.0];
        _botRightLabel.adjustsFontSizeToFitWidth = YES;
        _botRightLabel.font                      = [UIFont fontWithName:@"HelveticaNeue-Light" size:15.0];
        _botRightLabel.minimumScaleFactor        = 12.0/[_botRightLabel.font pointSize];
        _botRightLabel.textAlignment             = NSTextAlignmentRight;
        [self.contentView addSubview:_botRightLabel];


		_separatorView = [[UIView alloc] initWithFrame:CGRectZero];
		_separatorView.backgroundColor = [UIColor colorWithWhite:200.0/255.0 alpha:1.0];
		[self.contentView addSubview:_separatorView];

        self.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    }

    return self;
}


- (NSString *)accessibilityLabel
{
    NSString *label = [NSString stringWithFormat:@"%@, %@",
                            (_topLeftAccessibilityLabel) ? _topLeftAccessibilityLabel : _topLeftLabel.text,
                            (_botLeftAccessibilityLabel) ? _botLeftAccessibilityLabel : _botLeftLabel.text];

    if (cellState == UITableViewCellStateDefaultMask) {

        if (_topRightAccessibilityLabel)
            label = [label stringByAppendingFormat:@", %@", _topRightAccessibilityLabel];

        if (_botRightAccessibilityLabel)
            label = [label stringByAppendingFormat:@" %@",  _botRightAccessibilityLabel];
    }

    return label;
}


// Remember target state for transition
- (void)willTransitionToState:(UITableViewCellStateMask)newState
{
    [super willTransitionToState:newState];
    cellState = newState;
}


// Reset to default state before reuse of cell
- (void)prepareForReuse
{
    [super prepareForReuse];
    cellState = UITableViewCellStateDefaultMask;
}


- (void)layoutSubviews
{
    CGFloat margin = 15.0;

    // offset to compensate shift caused by editing control
    CGFloat editOffset = 0;

    if (cellState & UITableViewCellStateShowingEditControlMask)
        editOffset = 38.0;

    // space that can be distributed
    CGFloat width = self.frame.size.width - 9 - margin;

    // width of right labels
    CGFloat iWidth  = large ?  96.0 : 135.0;

    // y position and height of top right label
    CGFloat iYStart = large ?  17.0 :  20.0;
    CGFloat iHeight = large ?  36.0 :  30.0;
	CGFloat separatorHeight = 1.0 / [UIScreen mainScreen].scale;

    // compute label frames
    _topLeftLabel.frame  = CGRectMake(margin,                      20,       width - iWidth - 20, 30);
    _botLeftLabel.frame  = CGRectMake(margin,                      52,       width - iWidth - 20, 20);
    _topRightLabel.frame = CGRectMake(width - iWidth - editOffset, iYStart, iWidth - margin,      iHeight);
    _botRightLabel.frame = CGRectMake(width - iWidth - editOffset, 52,      iWidth - margin,      20);
	_separatorView.frame = CGRectMake(0, self.frame.size.height - separatorHeight, self.frame.size.width, separatorHeight);

    // hide right labels in editing modes
    [UIView animateWithDuration:0.5
                     animations:^{

                         CGFloat newAlpha = (cellState & UITableViewCellStateShowingEditControlMask) ? 0.0 : 1.0;

                         _topRightLabel.alpha = newAlpha;
                         _botRightLabel.alpha = newAlpha;
                     }];

    [super layoutSubviews];
}

@end

File: Classes/QuadInfoCell.swift
//
//  QuadInfoCell.swift
//  kraftstoff
//
//  Created by Ingmar Stein on 03.05.15.
//
// TableView cells with four labels for information.

import UIKit

final class QuadInfoCell: UITableViewCell {
  private(set) var topLeftLabel: UILabel
  private(set) var botLeftLabel: UILabel
  private(set) var topRightLabel: UILabel
  private(set) var botRightLabel: UILabel

  var topLeftAccessibilityLabel: String?
  var botLeftAccessibilityLabel: String?
  var topRightAccessibilityLabel: String?
  var botRightAccessibilityLabel: String?

  private var cellState: UITableViewCell.StateMask
  var large = false {
    didSet {
      topRightLabel.font = UIFont.preferredFont(forTextStyle: large ? UIFont.TextStyle.title1 : UIFont.TextStyle.title2)
    }
  }

  private func setupSubviews() {
    topLeftLabel.backgroundColor = .clear
    topLeftLabel.textColor = .label
    topLeftLabel.adjustsFontSizeToFitWidth = true
    topLeftLabel.translatesAutoresizingMaskIntoConstraints = false
    topLeftLabel.adjustsFontForContentSizeCategory = true
    topLeftLabel.font = UIFont.preferredFont(forTextStyle: UIFont.TextStyle.title1)
    contentView.addSubview(topLeftLabel)

    botLeftLabel.backgroundColor = .clear
    botLeftLabel.textColor = .highlightedText
    botLeftLabel.adjustsFontSizeToFitWidth = true
    botLeftLabel.translatesAutoresizingMaskIntoConstraints = false
    botLeftLabel.adjustsFontForContentSizeCategory = true
    botLeftLabel.font = UIFont.preferredFont(forTextStyle: UIFont.TextStyle.body)
    contentView.addSubview(botLeftLabel)

    topRightLabel.backgroundColor = .clear
    topRightLabel.textColor = .label
    topRightLabel.adjustsFontSizeToFitWidth = true
    topRightLabel.textAlignment = .right
    topRightLabel.translatesAutoresizingMaskIntoConstraints = false
    topRightLabel.adjustsFontForContentSizeCategory = true
    topRightLabel.font = UIFont.preferredFont(forTextStyle: large ? UIFont.TextStyle.title1 : UIFont.TextStyle.title2)
    contentView.addSubview(topRightLabel)

    botRightLabel.backgroundColor = .clear
    botRightLabel.textColor = .highlightedText
    botRightLabel.adjustsFontSizeToFitWidth = true
    botRightLabel.textAlignment = .right
    botRightLabel.translatesAutoresizingMaskIntoConstraints = false
    botRightLabel.adjustsFontForContentSizeCategory = true
    botRightLabel.font = UIFont.preferredFont(forTextStyle: UIFont.TextStyle.body)
    contentView.addSubview(botRightLabel)

    // setup constraints
    let views = ["topLeftLabel": topLeftLabel, "botLeftLabel": botLeftLabel, "topRightLabel": topRightLabel, "botRightLabel": botRightLabel]

    let constraints1 = [
      NSLayoutConstraint(item: topLeftLabel, attribute: .lastBaseline, relatedBy: .equal, toItem: topRightLabel, attribute: .lastBaseline, multiplier: 1.0, constant: 0.0),
      NSLayoutConstraint(item: botLeftLabel, attribute: .lastBaseline, relatedBy: .equal, toItem: botRightLabel, attribute: .lastBaseline, multiplier: 1.0, constant: 0.0),
      NSLayoutConstraint(item: topLeftLabel, attribute: .left, relatedBy: .equal, toItem: botLeftLabel, attribute: .left, multiplier: 1.0, constant: 0.0),
      NSLayoutConstraint(item: topRightLabel, attribute: .right, relatedBy: .equal, toItem: botRightLabel, attribute: .right, multiplier: 1.0, constant: 0.0),
    ]
    let constraints = constraints1
      + NSLayoutConstraint.constraints(withVisualFormat: "V:|-(20)-[topLeftLabel]-(2)-[botLeftLabel]-(20)-|", options: [], metrics: nil, views: views)
      + NSLayoutConstraint.constraints(withVisualFormat: "H:|-(15)-[topLeftLabel]-(2)-[topRightLabel]-(15)-|", options: [], metrics: nil, views: views)
    NSLayoutConstraint.activate(constraints)

    accessoryType = .disclosureIndicator
  }

  override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
    cellState = []
    botLeftLabel = UILabel(frame: .zero)
    topLeftLabel = UILabel(frame: .zero)
    topRightLabel = UILabel(frame: .zero)
    botRightLabel = UILabel(frame: .zero)

    super.init(style: style, reuseIdentifier: reuseIdentifier)

    setupSubviews()
  }

  required init?(coder aDecoder: NSCoder) {
    cellState = []
    botLeftLabel = UILabel(frame: .zero)
    topLeftLabel = UILabel(frame: .zero)
    topRightLabel = UILabel(frame: .zero)
    botRightLabel = UILabel(frame: .zero)

    super.init(coder: aDecoder)

    setupSubviews()
  }

  override var accessibilityLabel: String? {
    get {
      let topLeftText = (topLeftAccessibilityLabel ?? topLeftLabel.text) ?? ""
      let botLeftText = (botLeftAccessibilityLabel ?? botLeftLabel.text) ?? ""
      var label = "\(topLeftText), \(botLeftText))"
      if cellState == [] {
        if let accessibilityLabel = topRightAccessibilityLabel {
          label = "\(label), \(accessibilityLabel)"
        }

        if let accessibilityLabel = botRightAccessibilityLabel {
          label = "\(label) \(accessibilityLabel)"
        }
      }

      return label
    }
    set {}
  }

  // Remember target state for transition
  override func willTransition(to state: UITableViewCell.StateMask) {
    super.willTransition(to: state)
    cellState = state
  }

  // Reset to default state before reuse of cell
  override func prepareForReuse() {
    super.prepareForReuse()
    cellState = []
  }

  override func layoutSubviews() {
    super.layoutSubviews()

    // hide right labels in editing modes
    UIViewPropertyAnimator.runningPropertyAnimator(withDuration: 0.5, delay: 0, options: [], animations: {
      let newAlpha: CGFloat = self.cellState.contains(UITableViewCell.StateMask.showingEditControl) ? 0.0 : 1.0
      self.topRightLabel.alpha = newAlpha
      self.botRightLabel.alpha = newAlpha
    })
  }

  func reset() {
    topLeftLabel.text = ""
    topRightLabel.text = ""
    botLeftLabel.text = ""
    botRightLabel.text = ""
  }
}
